src/libxtp/basisset.cc

```diff
@@ -26,169 +26,101 @@
 namespace votca {
 namespace xtp {
 
-Index FindLmax(const std::string& type) {
-
+L StringToEnum(const std::string& type) {
   assert(!type.empty() && "Shelltype must be non empty!");
+  assert(type.size() == 1 && "Shelltype size must be one");
   const char t = type.back();
-  Index lmax = 0;
-  if (t == 'S') {
-    lmax = 0;
-  } else if (t == 'P') {
-    lmax = 1;
-  } else if (t == 'D') {
-    lmax = 2;
-  } else if (t == 'F') {
-    lmax = 3;
-  } else if (t == 'G') {
-    lmax = 4;
-  } else if (t == 'H') {
-    lmax = 5;
-  } else if (t == 'I') {
-    lmax = 6;
-  } else {
-    throw std::runtime_error("FindLmax: Shelltype not known");
-  }
-  return lmax;
+  return StringToEnum(t);
 }
 
-Index FindLmin(const std::string& type) {
-  assert(!type.empty() && "Shelltype must be non empty!");
-  Index lmin = std::numeric_limits<Index>::max();
-  const char t = type[0];
-  if (t == 'S') {
-    lmin = 0;
-  } else if (t == 'P') {
-    lmin = 1;
-  } else if (t == 'D') {
-    lmin = 2;
-  } else if (t == 'F') {
-    lmin = 3;
-  } else if (t == 'G') {
-    lmin = 4;
-  } else if (t == 'H') {
-    lmin = 5;
-  } else if (t == 'I') {
-    lmin = 6;
+L StringToEnum(char type) {
+  L l;
+  if (type == 'S') {
+    l = L::S;
+  } else if (type == 'P') {
+    l = L::P;
+  } else if (type == 'D') {
+    l = L::D;
+  } else if (type == 'F') {
+    l = L::F;
+  } else if (type == 'G') {
+    l = L::G;
+  } else if (type == 'H') {
+    l = L::H;
+  } else if (type == 'I') {
+    l = L::I;
   } else {
-    throw std::runtime_error("FindLmin: Shelltype not known");
+    throw std::runtime_error("FindLmax: Shelltype '" + std::string(1, type) +
+                             "' not known");
   }
-  return lmin;
+  return l;
 }
 
-Index OffsetFuncShell(const std::string& shell_type) {
-  Index nbf = std::numeric_limits<Index>::max();
-  assert(!shell_type.empty() && "Shelltype must be non empty!");
-  const char t = shell_type[0];
-  if (t == 'S') {
-    nbf = 0;
-  } else if (t == 'P') {
-    nbf = 1;
-  } else if (t == 'D') {
-    nbf = 4;
-  } else if (t == 'F') {
-    nbf = 9;
-  } else if (t == 'G') {
-    nbf = 16;
-  } else if (t == 'H') {
-    nbf = 25;
-  } else if (t == 'I') {
-    nbf = 36;
-  } else {
-    throw std::runtime_error("OffsetFuncShell: Shelltype not known");
+std::string EnumToString(L l) {
+  switch (l) {
+    case L::S:
+      return "S";
+    case L::P:
+      return "P";
+    case L::D:
+      return "D";
+    case L::F:
+      return "F";
+    case L::G:
+      return "G";
+    case L::H:
+      return "H";
+    case L::I:
+      return "I";
   }
-  return nbf;
+  return "";
```

Comment:
Why not raising an error here?

Reply:
If I am not mistaken you will get a compiler warning if you do not treat the whole enum so this cannot happen anymore.

Reply:
https://stackoverflow.com/questions/37254496/c-warning-enumeration-value-not-handled-in-switch-wswitch

Reply:
thanks, I didn't know!